subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CaptureStateAfterExecutionStepTest.groovy

```diff
@@ -105,7 +115,53 @@ class CaptureStateAfterExecutionStepTest extends StepSpec<BeforeExecutionContext
         0 * _
     }
 
-    // TODO Add tests for capturing overlapping outputs
+    def "overlapping outputs are captured"() {
+        def delegateDuration = Duration.ofMillis(123)
+
+        def staleFile = fileSnapshot("stale", HashCode.fromInt(123))
+        def outputFile = fileSnapshot("outputs", HashCode.fromInt(345))
+
+        def emptyDirectory = directorySnapshot()
+        def directoryWithStaleFile = directorySnapshot(staleFile)
+        def directoryWithStaleFileAndOutput = directorySnapshot(outputFile, staleFile)
+        def directoryWithOutputFile = directorySnapshot(outputFile)
+
+        def previousOutputs = ImmutableSortedMap.<String, FileSystemSnapshot> of(
+            "outputDir", emptyDirectory
+        )
+        def outputsBeforeExecution = ImmutableSortedMap.<String, FileSystemSnapshot> of(
+            "outputDir", directoryWithStaleFile
+        )
+        def outputsAfterExecution = ImmutableSortedMap.<String, FileSystemSnapshot> of(
+            "outputDir", directoryWithStaleFileAndOutput
+        )
+        def filteredOutputs = ImmutableSortedMap.<String, FileSystemSnapshot> of(
+            "outputDir", directoryWithOutputFile
+        )
+        def overlappingOutputs = Mock(OverlappingOutputs)
+
+        when:
+        def result = step.execute(work, context)
+        then:
+        result.afterExecutionState.get().outputFilesProducedByWork == filteredOutputs
+        !result.reused
+        result.duration == delegateDuration
+        result.originMetadata.buildInvocationId == buildInvocationScopeId.asString()
+        result.originMetadata.executionTime >= result.duration
+
+        1 * delegate.execute(work, context) >> delegateResult
+        1 * delegateResult.duration >> delegateDuration
+        _ * context.beforeExecutionState >> Optional.of(Mock(BeforeExecutionState) {
+            _ * detectedOverlappingOutputs >> Optional.of(overlappingOutputs)
+            _ * outputFileLocationSnapshots >> outputsBeforeExecution
+        })
+        _ * context.previousExecutionState >> Optional.of(Mock(PreviousExecutionState) {
+            _ * outputFilesProducedByWork >> previousOutputs
+        })
```

Comment:
Should these two be stubs, so that we could lose the `_ *`?

Reply:
Good point! I replaced the mocks by stubs.